lib/addons/argocd/index.ts

```diff
@@ -1,26 +1,359 @@
+import { HelmChart, KubernetesManifest, ServiceAccount } from "@aws-cdk/aws-eks";
+import { ManagedPolicy } from "@aws-cdk/aws-iam";
+import { SecretsManager } from "aws-sdk";
+
 import { ClusterAddOn, ClusterInfo, ClusterPostDeploy } from "../../stacks/cluster-types";
 import { Team } from "../../teams";
+import * as yaml from 'yaml';
+import { btoa } from '../../utils/string-utils';
+import { Construct } from "@aws-cdk/core";
+
+import bcrypt = require('bcrypt');
+
+
+export interface ArgoApplicationRepository {
+    /**
+     * Expected to support helm style repo at the moment
+     */
+    repoUrl: string,
+
+    /** 
+     * Path within the repository 
+     */
+    path?: string,
+
+    /**
+     * Optional name for the bootstrap application
+     */
+    name?: string,
+
+    /**
+     * Secret from AWS Secrets Manager to import credentials to access the specified git repository.
+     * The secret must exist in the same region and account where the stack will run. 
+     */
+    credentialsSecretName?: string,
+
+    /**
+     * Depending on credentials type the arn should either point to an SSH key (plain text value)
+     * or a json file with username/password attributes.
+     * For TOKEN type per ArgoCD documentation (https://argoproj.github.io/argo-cd/user-guide/private-repositories/) 
+     * username can be any non-empty username and token value as password.
+     */
+    credentialsType?: "USERNAME" | "TOKEN" | "SSH"
+
+}
+
+/**
+ * Configuration options for ArgoCD add-on.
+ */
+export interface ArgoCDAddOnProps {
+
+    /**
+     * Namespace where add-on will be deployed. 
+     */
+    namespace?: string,
+
+    /**
+     * If provided, the addon will bootstrap the app or apps in the provided repository.
+     * In general, the repo is expected to have the app of apps, which can enable to bootstrap all workloads,
+     * after the infrastructure and team provisioning is complete. 
+     */
+    bootstrapRepo?: ArgoApplicationRepository,
+
+    /**
+     * Optional admin password secret (plaintext).
+     * This allows to control admin password across the enterprise. Password will be retrieved and 
+     * store as bcrypt hash. 
+     * Note: at present, change of password will require manual restart of argocd server. 
+     */
+    adminPasswordSecretName?: string,
 
+    /**
+     * Values to pass to the chart as per https://github.com/argoproj/argo-helm/blob/master/charts/argo-cd/values.yaml.
+     */
+    values?: {
+        [key: string]: any;
+    };
+}
+
+
+const argoDefaults: ArgoCDAddOnProps = {
+    namespace: "argocd"
+}
+
+/**
+ * Local function to create a secret reference for SSH key.
+ * @param url 
+ * @param secretName 
+ * @returns 
+ */
+const sshRepoRef = (url: string, secretName: string): string => yaml.stringify(
```

Comment:
move these to utils file in the same dir?

Reply:
sure